### ISSUE TITLE
baton-metaquery avu filter: absence of an attribute ?

### ISSUE THREAD

**Issue by mcast** [wtsi-npg/baton#105]
I would like to list files (owned by a user and...) _not_ having `avus: [{a:"id_file"}]` attached.  My reading of the 0.12.1 docs and subsequent experiments suggest it isn't supported.

I'm importing legacy files and I need to ensure that the `ireg` has been followed by the appropriate `baton-metamod`, to supply legacy primary key and other metadata.

There are enough files that I'm doubting I can fetch the entire list back in one bite, then subtract the ones which do have `id_file`, but I will try it.

Possible workaround for me: import to a different collection, then `imv` into the right place after metamod?

Suggestions gratefully received,

Matthew

**Comment by mcast** [comment]
> Possible workaround for me: import to a different collection,
> then imv into the right place after metamod?

It turns out I need to do this anyway, because `imv` is somewhat faster than `ireg` per file.
I guess the attribute-(non-)presence test could still be useful though.

**Comment by keithj** [comment]
The iRODS query API is not expressive enough to allow a 'not' query.

As you've found, ````imv```` retains metadata (unless you move items across zones!), so allows you to stage files in one place and then move them to another. In case you are wondering , you can't ensure that the metadata are intact in the ACID sense. The iRODS server autocommits every AVU modification while you're working and the results become visible to other users' queries.

**Comment by mcast** [comment]
On Fri, Mar 06, 2015 at 05:43:04AM -0800, Keith James wrote:

> The iRODS query API is not expressive enough to allow a 'not' query.

OK, bang goes that one; but it wasn't going to prop up an illusion of
transactions anyway.

(If Baton wanted to provide the illusion of 'no attribute' queries it
would be more efficient to do the set-difference without streaming
both sets to JSON and back.  I don't need it.)


> As you've found, ````imv```` retains metadata (unless you move items
> across zones!), so allows you to stage files in one place and then
> move them to another.

Yes.  Next question is whether I can parallelise the `imv`s, I'm
getting about 15,000/hour...  from a remote client.

A `baton-imv` could be a useful tool to get around this, but I should
be fine without it.


> In case you are wondering, you can't ensure that the metadata are
> intact in the ACID sense.  The iRODS server autocommits every AVU
> modification while you're working and the results become visible to
> other users' queries.

This is the first time I've seen it written, but I was Beginning To
Suspect.  (For the record - irods 3.3.1 here.)


Thanks for explaining.  Close ticket?
-- 
Matthew


-- 
 The Wellcome Trust Sanger Institute is operated by Genome Research 
 Limited, a charity registered in England with number 1021457 and a 
 company registered in England with number 2742969, whose registered 
 office is 215 Euston Road, London, NW1 2BE.

**Comment by keithj** [comment]
Another option is ````iput````, which has a parallel mode for inserting a directory hierarchy.

**Comment by mcast** [comment]
I re-worked with `imv` and this is a better approach - it's still slightly faster than individual `ireg` per file, even with setting AVUs.

If you happen to know whether running `baton-metasuper ...; imv ...` in parallel streams will produce a speedup, this would be a good time to let me know.  :smile:

Also a multi-file or multi-action subcommand to do the above on several files in sequence would presumably save more overhead, but I'm not going to wait for that.

I can't `iput` yet because a) there is no space yet b) it would take at least 11 weeks divided by parallelism and c) it would reduce some back-compatibility that will ease our migration.


Other than that...  nothing left for this issue, thanks for your advice!

Matthew

**Comment by keithj** [comment]
If you want the best speed with baton operations, it's best to pipeline all metadata operations through a single client if you can. That way you avoid the overhead of starting new processes and  opening new connections to the server (this use case is why baton was written). You will get speed through parallel processes too, but watch out that you don't exhaust the maximum number of database connections. AFAIK, one client == one database connection between the iRODS server and the backend database i.e. no pooling.

**Comment by mcast** [comment]
On Mon, Mar 09, 2015 at 08:28:40AM -0700, Keith James wrote:

> If you want the best speed with baton operations, it's best to
> pipeline all metadata operations through a single client if you can.

It wasn't clear to me from the docs that I could re-use the connection
for more than just all the keys on one file.  I had a poke at it and
it went bang again - I made a separate issue (#107).


> That way you avoid the overhead of starting new processes and
> opening new connections to the server (this use case is why baton
> was written). You will get speed through parallel processes too, but
> watch out that you don't exhaust the maximum number of database
> connections. AFAIK, one client == one database connection between
> the iRODS server and the backend database i.e. no pooling.

This all makes sense.  While I have to call `imv`, I can only
asymptotically approach 15,000 files/hour anyway.

-- 
Matthew


-- 
 The Wellcome Trust Sanger Institute is operated by Genome Research 
 Limited, a charity registered in England with number 1021457 and a 
 company registered in England with number 2742969, whose registered 
 office is 215 Euston Road, London, NW1 2BE.
